### ISSUE TITLE
Replicating cross-encoder/ms-marco-MiniLM-L12-v2

### ISSUE THREAD

**Issue by MathVast** [UKPLab/sentence-transformers#3566]
Hello everyone! Thanks a lot for the great work and library!

I have been trying to reproduce the fine-tuning of the MiniLM-L12-v2 cross-encoder model, starting from the pretrained checkpoint of "microsoft/MiniLM-L12-H384-uncased". After investigating to find resources, I felt upon @tomaarsen [recipe](https://huggingface.co/tomaarsen/reranker-MiniLM-L12-H384-uncased-msmarco-bce/blob/main/README.md?utm_source=chatgpt.com) for training an equivalent model, starting from the same pretrained model. I thought this would be appropriate, given the indications provided in the [sentence-transformers repo](https://github.com/huggingface/sentence-transformers/blob/master/examples/cross_encoder/training/ms_marco/README.md). 

Unfortunately, so far, all my attempts have failed and I end up with a model that either overfits on MSMARCO and is pretty bad on other BEIR datasets or, that can't match the reported numbers for any of the datasets I evaluate it on.

Digging further, I felt upon these 2 issues: https://github.com/huggingface/sentence-transformers/issues/1058 and https://github.com/huggingface/sentence-transformers/issues/1669, that lead me to think that perhaps BCE is not the way to go and I should instead switch to knowledge distillation.

Can anyone give me some pointers at what I'm missing and why I can't reproduce the fine-tuned model "cross-encoder/ms-marco-MiniLM-L12-v2"?

Thanks a lot in advance!

**Comment by tomaarsen** [comment]
Hello!

Apologies for the delay, I was on a brief vacation when you opened this.

Nils Reimers trained the https://huggingface.co/cross-encoder/ms-marco-MiniLM-L12-v2 model, and he linked to this page for his source code: https://github.com/huggingface/sentence-transformers/tree/main/examples/training/ms_marco
This doesn't exist anymore, but 5 days before uploading the model, he pushed c66688b9fb05b6e36bf543315670ac62d420601b and 9f76b5790fa8f0175ccdc27e41f9ae348f557144, introducing the `train_cross-encoder-v2.py` script: https://github.com/huggingface/sentence-transformers/blob/9f76b5790fa8f0175ccdc27e41f9ae348f557144/examples/training/ms_marco/train_cross-encoder-v2.py

This script finetunes a CrossEncoder based on `microsoft/MiniLM-L12-H384-uncased` using [Hofstätter et al.](https://arxiv.org/abs/2010.02666) ensemble data, using knowledge distillation (specifically, the MSE Loss). 

During the v4 release of Sentence Transformers a few years later, I separated the cross-encoder and sentence-transformers training examples, and modernized the `model.fit`-style training from the script into a Trainer-style training. The file was replaced with https://github.com/huggingface/sentence-transformers/blob/main/examples/cross_encoder/training/distillation/train_cross_encoder_kd_mse.py, which was designed to match the behaviour of the old script, bar some slight differences. At a glance: 
* The old script used a learning rate of 7e-6, while the new one uses 8e-6.
* The old script uses a batch size of 32, while the new one uses 16.
* The old script limits the model's max. sequence length to 512 and sets the activation function (i.e. score postprocessing) to Identity rather than Sigmoid which is the default for the new script.
* The old script uses 5000 steps for warmup, and the new script uses a warmup ratio of 0.1
* The old script uses `use_amp` (a.k.a. fp16), while the new script uses bf16.
* The new script uses `load_best_model_at_end=True` to grab the best checkpoint. Rerankers often overfit pretty early. Nils probably also used the best checkpoint in his work.
* Perhaps most of all: the new script limits the dataset to 2 million samples, whereas the old one uses the full dataset by Hofstätter et al. I believe the full dataset has 80 million pairs.

I reuploaded the Hofstätter data here: https://huggingface.co/datasets/sentence-transformers/msmarco/viewer/bert-ensemble-mse?views%5B%5D=bert_ensemble_mse, so that the new training script can use it. I believe that this model was produced by my new training script: https://huggingface.co/tomaarsen/reranker-MiniLM-L12-H384-msmarco-mse, although it seems that I used a slightly higher batch size (64) than the new script on the GitHub. You can inspect the metrics, as well as the training logs over time. Interestingly, there's no notable overfitting here, with the last checkpoint being the chosen one.

This model scores:
* NanoMSMARCO NDCG@10: 0.6585 (+0.1181)
* NanoNFCorpus NDCG@10: 0.3864 (+0.0613)
* NanoNQ NDCG@10: 0.7366 (+0.2359)

The section in brackets denotes the improvement on top of a BM25 first-stage retriever.

When I also run the NanoBEIR benchmark with https://huggingface.co/cross-encoder/ms-marco-MiniLM-L12-v2, that model scores:
* NanoMSMARCO NDCG@10: 0.6746 (+0.1342)
* NanoNFCorpus NDCG@10: 0.3994 (+0.0743)
* NanoNQ NDCG@10: 0.7825 (+0.2818)

As you can see, the cross-encoder model is certainly stronger, but the new script trained with slightly different parameters and 1/40th the data does get close. 

In short, if you want to reproduce the cross-encoder model: try https://github.com/huggingface/sentence-transformers/blob/main/examples/cross_encoder/training/distillation/train_cross_encoder_kd_mse.py and consider modifications to match https://github.com/huggingface/sentence-transformers/blob/9f76b5790fa8f0175ccdc27e41f9ae348f557144/examples/training/ms_marco/train_cross-encoder-v2.py.

- Tom Aarsen

**Comment by MathVast** [comment]
Hello, thanks a lot for the very complete answer!

No worries about the delay, it left me some time to dig further and identify some limitations in my own training script. In the end, I converged to this script: https://github.com/huggingface/sentence-transformers/blob/main/examples/cross_encoder/training/distillation/train_cross_encoder_kd_margin_mse.py which uses the MarginMSE loss instead of the MSE. I tried both (though not on the 80M pairs but only on the limited 2M setup) and MarginMSE seems to produce better results. 

I also copied the hyper-parameters from this script, which are similar to the ones in https://github.com/huggingface/sentence-transformers/blob/main/examples/cross_encoder/training/distillation/train_cross_encoder_kd_mse.py , so I'll definitely try to update them with https://github.com/huggingface/sentence-transformers/blob/9f76b5790fa8f0175ccdc27e41f9ae348f557144/examples/training/ms_marco/train_cross-encoder-v2.py .

Finally, as I was still unable to reproduce https://huggingface.co/cross-encoder/ms-marco-MiniLM-L12-v2 even after all these changes, I stumbled upon an answer from Nils to an old issue, where he pointed out the impact of the randomness induced by the GPU. To account for it, he suggested that a good method is to train k models (k=10 for example) and compare their averaged performance instead. I don't know what's your opinion on this and if you think this could also be a part of the problem.

Thanks again for the answer, hope I'll be able to make some progress on this matter with these additional inputs!

**Comment by tomaarsen** [comment]
Awesome, nice work! I'm also currently retraining with the full 80M to see what that'll do. I also want to try it with the excellent Ettin base models.

I also agree that MarginMSE should outperform MSE.

> To account for it, he suggested that a good method is to train k models (k=10 for example) and compare their averaged performance instead. I don't know what's your opinion on this and if you think this could also be a part of the problem.

This sounds like seed optimization: https://github.com/huggingface/sentence-transformers/blob/main/examples/sentence_transformer/training/data_augmentation/train_sts_seed_optimization.py
In short, the paper shows that minor variations in how data is sampled etc. can affect model training to a surprisingly large extent. Retraining with a different seed can then help with performance. Personally, I think the newer, modern version of this is model souping: merging different checkpoints (either with different training setups, or just different seeds) to outperform each of the individual checkpoints.

- Tom Aarsen

**Comment by MathVast** [comment]
Cool, I'm eager to see what you'll be able to get out of the Ettin models!

Thanks for the pointer, indeed, model souping was one of the way I imagined I could value the multiple models I'd end up with after the different trainings. In the first place though, it was really just to limit the impact of randomness on my results and be able to draw meaningful comparisons, rather than finding the best model possible.

**Comment by tomaarsen** [comment]
Here's my Ettin models trained using MarginMSE using the full dataset: https://huggingface.co/collections/tomaarsen/strong-and-small-rerankers

* https://huggingface.co/tomaarsen/ms-marco-ettin-17m-reranker: 0.5764 NDCG@10 on 3 datasets from NanoBEIR (MS MARCO, NFCorpus, NQ)
* https://huggingface.co/tomaarsen/ms-marco-ettin-32m-reranker: 0.6128 NDCG@10 ...
* https://huggingface.co/tomaarsen/ms-marco-ettin-68m-reranker: 0.6355 NDCG@10 ...
* https://huggingface.co/tomaarsen/ms-marco-ettin-150m-reranker: 0.6548 NDCG@10 ...

Some baselines:
* https://huggingface.co/BAAI/bge-reranker-base (~300M): 0.5592 NDCG@10 ...
* https://huggingface.co/BAAI/bge-reranker-large (~600M): 0.5504 NDCG@10 ...
* https://huggingface.co/BAAI/bge-reranker-v2-m3 (~600M): 0.6184 NDCG@10 ...
* https://huggingface.co/ibm-granite/granite-embedding-reranker-english-r2 (149M): 0.6306 NDCG@10 ...
* https://huggingface.co/cross-encoder/ms-marco-MiniLM-L12-v2 (33M): 0.6188 NDCG@10 ...
* https://huggingface.co/cross-encoder/ms-marco-MiniLM-L6-v2 (22M): 0.6071 NDCG@10 ...

Note that this benchmark is not the be-all-end-all, especially because I trained on the train split from MS MARCO, while evaluating against the eval split from MS MARCO (among others). It does hint that it's possible to train SOTA models for whatever domain you want. The trainings took 1.5 to 5 hours for 17m to 150m, but granted, I used excessive hardware as you can see in the model card. 

I also reached approximately 0.6188 on my MiniLM-L12-H384-uncased finetune, reproducing the `cross-encoder/ms-marco-MiniLM-L12-v2` model.

- Tom Aarsen

**Comment by MathVast** [comment]
Woah the Ettin models look competitive. I guess you did little to no hyperparameter search on top of that did you?

> Note that this benchmark is not the be-all-end-all

What do you mean by that?

For the MiniLM based model, do you know which step corresponds to the best version of the model on your side? I am currently evaluating the different checkpoints of my own reproduction script and so far, I don't really see any additional value in training on more data points than the initial 2M from the original scripts I copied.

**Comment by tomaarsen** [comment]
> I guess you did little to no hyperparameter search on top of that did you?

No HPO indeed, I just increased the learning rate from 7e-6 to 2e-5 as I'm training with much larger batch sizes. 

> What do you mean by that?

The benchmark I chose (the NanoBEIR with 3 datasets) is a bit arbitrary, and the ranking on that benchmark might not correspond with the ranking of other domains/tasks. For example, the BAAI models were particularly bad on NQ and pretty good on the other 2. 

<img width="1541" height="732" alt="Image" src="https://github.com/user-attachments/assets/00688d62-2d29-49bc-9a77-cb5c08742ce7" />

Here are my evaluation results across the 40m MarginMSE triplets. As you can see, except for the 150m, the scores do keep growing, but this is likely also due to the learning rate scheduling. I suspect more data would only help very slightly. With BinaryCrossEntropyloss I usually start overfitting much earlier, and then picking an early checkpoint is often better.

- Tom Aarsen

**Comment by MathVast** [comment]
> The benchmark I chose (the NanoBEIR with 3 datasets) is a bit arbitrary, and the ranking on that benchmark might not correspond with the ranking of other domains/tasks. For example, the BAAI models were particularly bad on NQ and pretty good on the other 2.

Okay, I agree. I did something similar for my tests and picked smaller "minified" versions of 4 BEIR datasets. For the validation though, I just use a subset of topics from the validation set of MSMARCO, which may also explains why I struggle finding a suitable checkpoint that can compete with `cross-encoder/ms-marco-MiniLM-L12-v2`.

I note that MSE seems to also give better results than MarginMSE, interesting. You didn't compared the two losses over the Ettin models?

**Comment by tomaarsen** [comment]
> I note that MSE seems to also give better results than MarginMSE, interesting. You didn't compared the two losses over the Ettin models?

No, MarginMSE should be stronger on paper, and I figured the difference was due to the chosen learning rate being more suitable for MSE. So I only tested MarginMSE when I moved to trying the Ettin models.

- Tom Aarsen

**Comment by MathVast** [comment]
Well, thank you for exploring all of this and for answering my initial request with so many details. It really helped me. As I should be able to move forward now, I think we can close this issue now.

Again, thanks a lot and take care!
